AWS Route 53 GSLB Multi-Region Proxy Peering High Availability Deployment Guide

WilliamLoy opened the pull request:
This guide gives a high-level overview of deploying a single cluster available across multiple regions without using trusted clusters.

This architecture uses AWS Route53 Latency Routing to create Global Server Load Balancing for multi-region single cluster deployments. It also uses Proxy Peering to restrict the number of connections necessary to facilitate this Teleport Cluster architecture.

WilliamLoy committed:
Create aws-gslb-proxy-peering-ha-deployment.mdx

migrating to branch off main repo

WilliamLoy committed:
adding chart for new guide

WilliamLoy committed:
adding image for new guide

WilliamLoy committed:
Delete aws-gslb-proxy-peering-ha-deployment.png

zmb3 commented on docs/img/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.png:
teleport.com is not a domain that we own. I would update this image to use a domain we own (goteleport.com or teleport.sh, for example), or an obvious example domain like teleport.example.com.

WilliamLoy replied:
I updated the diagram to use goteleport.com

zmb3 commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,306 @@
+---
+title: "AWS Route 53 GSLB Multi-Region Proxy Peering High Availability Deployment Guide"
```
This title has a lot of jargon in it. Can we simplify it a bit?

WilliamLoy replied:
Would "AWS Multi-Region Single Cluster Deployment" work? I don't know if we need HA explicitly stated in the title.

webvictim replied:
I think having HA in the title is important.

AWS Multi-Region High Availability Deployment Guide

WilliamLoy replied:
Updated title

zmb3 commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,306 @@
+
+This deployment architecture isn’t recommended for use cases where your users or resources are 
+clustered in a single region or for Managed Service Providers needing to provide separate clusters 
+to customers. Additionally, this architecture is not a solution for increasing the scalability of 
```
This sentence is a little confusing.

- First we say "it's imperative to ensure the scalability"
- Then we say "this deployment architecture makes all resources accessible through a single Teleport Cluster"
- Lastly, we say "this is not a solution for increasing the scalability of a single cluster"

Feels a bit like rock paper scissors :-)

WilliamLoy replied:
I think removing "Additionally, this architecture is not a solution for increasing the scalability of 
a single cluster." clears up the confusion. I don't think we need to point out everything this architecture _doesn't do_.

zmb3 commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,306 @@
+This deployment architecure makes all connected resources accessible through a single Teleport cluster 
+across multiple regions using exclusively AWS ecosystem infrastructure. 
+
+This is accomplished using AWS Route 53 to create Global Server Load Balancing (GSLB) 
```
I might rephrase this a bit to make it clear what these two pieces provide. Maybe something like this:

This deployment architecture features two important design decisions:
1. AWS Route 53 is used for Global Server Load Balancing (GSLB). This allows for efficient distribution of traffic across resources that are globally distributed.
2. Teleport's Proxy Peering is used to reduce the total number of tunnel connections in the Teleport cluster.

WilliamLoy replied:
I like your recommendations. The only change I made was specifying Latency based routing is being used as I think readers could easily assume Geolocation routing is being used. It is worth repeating through the guide in my opinion because Latency based is going to improve performance whereas Geolocation won't (as much).

zmb3 commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,306 @@
+
+## Overview
+This deployment architecure makes all connected resources accessible through a single Teleport cluster 
+across multiple regions using exclusively AWS ecosystem infrastructure. 
```
I might move the note about "exclusively AWS ecosystem" down under the _Key deployment components_ section, since that's where you bring up all the AWS components.

zmb3 commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,306 @@
+- When Teleport Auth servers are limited to a single region, there is a higher likelihood 
+  of decreased availability during an AWS regional outage.
+- Technically complex to deploy
+- Long-term cost may be a prohibitive factor for some organizations and can increase total 
```
Can we elaborate on this? What is the highest contributor to the long-term cost? Are there alternatives?

WilliamLoy replied:
Infrastructure hosting cost probably isn't something we need to be giving opinions on, so I'm just going to remove this rather than elaborate.

zmb3 commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,306 @@
+architecture](../../img/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.png)
+
+
+## AWS Network load balancer(NLB)
```
```suggestion
## AWS Network load balancer (NLB)
```

zmb3 commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,306 @@
+
+## AWS Network load balancer(NLB)
+For this deployment architecture we recommend using AWS NLBs if you plan 
+to use Teleport TLS routing and Proxy Peering. The NLB forwards traffic 
```
We already listed that proxy peering is a key component of this architecture, so are NLBs a recommendation or required?

I think people looking for opinionated architectures like this want you to be very specific.

WilliamLoy replied:
"AWS NLBs are required for this highly available deployment architecture. 
The NLB forwards traffic from users and services to an available Teleport instance. This must not 
terminate TLS, and must transparently forward the TCP traffic it receives. 
In other words, this must be a Layer 4 load balancer, not a Layer 7 
(e.g., HTTP) load balancer. "

The NLBs aren't required for either of those components, so I removed the mention and pointed out that the NLBs are required for high availability in this deployment.

zmb3 commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,306 @@
+  title="Note"
+>
+We recommend enabling cross-zone load balancing for the Auth and Proxy service NLB configurations to route 
+traffic across multiple zones.Doing this improves resiliency against localized AWS zone outages.
```
```suggestion
traffic across multiple zones. Doing this improves resiliency against localized AWS zone outages.
```

zmb3 reviewed (commented):
Just a start, only got halfway through.

WilliamLoy committed:
Addressing reviewer comments

I applied changes based on @zmb3's feedback. I also made changes to sections that were borrowed from other guides that conflict with this guide.

WilliamLoy commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,286 @@
+ensure that users can continue to access infrastructure should an outage or 
+incident affect the availability of your Teleport cluster. 
+
+In order to maintain optimal end-user 
+experience with minimal latency and maximum performance, it is imperative to ensure the scalability 
+of your Auth Service and Proxy Service to accommodate increasing numbers of users and connected resources.
```
I'll reword this to account for the Dynamo limits.

WilliamLoy replied:
I removed this paragraph and the one before it. They don't add useful information to the guide.

WilliamLoy commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,286 @@
+  type="warning"
+  title="Note"
+>
+We recommend enabling cross-zone load balancing for the Auth and Proxy service NLB configurations to route 
+traffic across multiple zones. Doing this improves resiliency against localized AWS zone outages.
```
Recommendation.

WilliamLoy replied:
If you disable cross-zone and have instances in zones outside of the same zone as the NLB it won't route the traffic properly.

WilliamLoy replied:
We would have to explain how to get that to work but I see your point. I'll explain that it is a requirement for this setup.

WilliamLoy commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,286 @@
+
+In this deployment architecure, Proxy Peering is used to restrict the number of connections made from 
+resources to proxies in the Teleport Cluster. Full Proxy Peering explanation and configuration details
+can be reviewed in the [Proxy Peering RFD](https://github.com/gravitational/teleport/blob/master/rfd/0069-proxy-peering.md).
```
I think that change makes sense.

WilliamLoy commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,286 @@
+you could incur additional latency because peer proxies must continually dial through the NLB until 
+they establish connection to the correct peer target.
+
+When using an NLB for peer_public_addr, be sure to set agent_connection_count to a value >=2.
```
I assumed the default would be sufficient. If it isn't, we should probably change the default setting.

dboslee replied:
Why does the agent_connection_count matter for NLBs?

WilliamLoy replied:
It is my understanding, based on the proxy peering RFD, that leaving agent_connection_count=1 increases the likelihood of agent unavailability. To your point, the recommendations don't necessarily have anything to do with the NLB, as the sentence would imply.
As @rosstimothy we should probably recommend >=2 agent_connection_count for option one as well.

WilliamLoy commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
````diff
@@ -0,0 +1,286 @@
+  ...
+```
+
+**Option 2:** Proxies can use peer_public_addr: to advertise the proxy NLB. When using this method 
+you could incur additional latency because peer proxies must continually dial through the NLB until 
+they establish connection to the correct peer target.
````
I'll add additional warning messages to option 2.

ptgott commented:
@WilliamLoy I know this is going through engineering review now, so let me know with an @ if you'd like me to give this a look for copy. I'll unsubscribe in the meantime to keep notifications low. Thanks!

dboslee commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,286 @@
+Highly available Teleport deployments use an object storage service for
+persisting session recordings.
+
+Depeding on how you choose to deploy the Teleport cluster, the Teleport Auth Service can create the required S3 bucket if it doens't already exist.
```
```suggestion
Depending on how you choose to deploy the Teleport cluster, the Teleport Auth Service can create the required S3 bucket if it doesn't already exist.
```

dboslee commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,286 @@
+
+<Admonition title="Required permissions">
+
+The Auth Service instances needs permissions to read from and write to DynamoDB, as well as 
```
Add link to dynamodb guide https://goteleport.com/docs/reference/backends/#dynamodb

dboslee commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,286 @@
+
+<Admonition title="Required permissions">
+
+The Auth Service instances need permissions to get S3 buckets as well as to create, get, list, 
```
Add link to s3 guide https://goteleport.com/docs/reference/backends/#s3

WilliamLoy committed:
Update aws-gslb-proxy-peering-ha-deployment.mdx

Changes based on @rosstimothy's recommendations.

WilliamLoy committed:
Update docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx

Co-authored-by: David Boslee <david@goteleport.com>

WilliamLoy committed:
updated architecture diagram

WilliamLoy committed:
Update aws-gslb-proxy-peering-ha-deployment.mdx

- corrected agent_connection_count guidance
- clarified cross-zone LB guidance
- added S3 and Dynamo guidance links
- Updated title

dboslee commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,282 @@
+
+## Global Server Load Balancing with Route 53
+
+[Latency-based routing](https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/routing-policy-latency.html) 
+in a private hosted zone must be used to ensure traffic from Teleport 
+resources are routed to the closest or lowest latency path Proxy NLB based on the region of 
+the VPC the resource is connecting from.
```
Does this assume the resources being connected are also hosted on AWS? How do users resolve DNS if its a private hosted zone?

WilliamLoy replied:
That is a mistake. I'll update it to reflect that the zone should be public.

WilliamLoy replied:
It's probably worth mentioning that route53 latency-based routing is only optimized to route traffic to AWS data centers so any resource outside AWS could have varied routing performance compared to the AWS resources.

dboslee replied:
As long as route53 is routing to NLBs as this guide suggest it shouldn't be an issue.

dboslee commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
````diff
@@ -0,0 +1,282 @@
+
+Proxies must advertise a peer address which can be configured to use one of the two options listed below:
+
+**Option 1:** You can set peer_public_addr: to the specific name of that proxy. This is the recommended 
+method for lowest latency and most reliable connection.
+
+```
+proxy_service:
+  ...
+  peer_public_addr: teleport.example.com:3021
+  agent_connection_count: 2
+  ...
+```
````
We should add a note that VPC peering is required for this configuration to work.

WilliamLoy replied:
I added a note about this in the Configure the Auth Service NLB section but I can replicate it in the proxy peering section as well.

WilliamLoy replied:
@dboslee would Transit Gateways not work here as well?

dboslee replied:
I haven't used it but it seems to be a similar service to VPC peering so I am sure it would work.

WilliamLoy commented on docs/img/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.png:
@rosstimothy I actually copied this from the RFD diagram because I thought it would be less confusing but I can revert it to just say Teleport Proxy.

dboslee replied:
sorry missed this. I agree it could be confusing as someone could interpret them as different services where in reality all proxies can handle both user and agent connections.

WilliamLoy replied:
I reverted it back to just teleport proxies

WilliamLoy commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,282 @@
+- Provides a highly-resilient, redundant HA architecture for Teleport that can quickly 
+  scale with an organization’s needs.
+- All required Teleport components can be provisioned within the AWS ecosystem.
+- Using load balancers for the Proxy and Auth services allows for increased availability 
+  during Teleport Cluster upgrades. Instances can easily be removed and added while 
+  limiting impact to active users.
```
Fair enough. I'll remove the second line and add a link to upgrading the cluster.

WilliamLoy commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,282 @@
+| - | - |
+| `3025` | TLS port used by the Auth Service to serve its API to other Nodes in a cluster |
+
+## Cluster state backend
```
I'm fine with that.

WilliamLoy commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,282 @@
+- Proxies are deployed across multiple regions
+- [AWS Route 53 latency based routing](https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/routing-policy-latency.html) 
+- [GSLB](https://www.cloudflare.com/learning/cdn/glossary/global-server-load-balancing-gslb/)
+- [Teleport TLS Routing](https://goteleport.com/docs/architecture/tls-routing/) to reduce the number of ports needed to use Teleport
```
@rosstimothy the ALPN headers are required to differentiate legacy agents from proxy peering agents, so TLS routing would be required for that contingency. 

I don't know if this impacts DNS latency routing.

dboslee replied:
ALPN can be used by intermediate load balancers (we do this on cloud in order to support tenants with proxy peering enabled or disabled or legacy agents) but the teleport proxy themselves don't handle the agent connections differently based on the ALPN. So TLS routing is not required but a proxy peering deployment requires v10+ agents to work as expected without more complex load balancing.

WilliamLoy replied:
@rosstimothy I think the primary use case is to use TLS Routing.

WilliamLoy replied:
@dboslee I played with some phrasing to try to place this and it felt like an unuseful line without complete instructions for setting up the cluster for non-TLS routing. Can we fold this into another guide or a v2 of this guide?

dboslee replied:
Yep I think its fine to leave out for now

WilliamLoy commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,282 @@
+provision Teleport Proxy Service instances with these credentials and renew them
+periodically. 
+
+For high-availability deployments that use Let's Encrypt to supply TLS
+credentials to Teleport instances running behind a load balancer, you need
+to use the [ACME
+DNS-01](https://letsencrypt.org/docs/challenge-types/#dns-01-challenge)
+challenge to demonstrate domain name ownership to Let's Encrypt. In this
+challenge, your TLS credential provisioning system creates a DNS TXT record with
+a value expected by Let's Encrypt.
```
@WilliamLoy LetsEncrypt only allows DNS-01 for wildcard cert requests which would be needed for app access. 

I can add a line explaining this if you think it is needed.

WilliamLoy commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,282 @@
+teleport:
+    nodename: ssh-node
+    ...
+    proxy_server: teleport.example.com:443
```
Good catch. No it will not work. I'll add the version number to the example.

WilliamLoy commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,282 @@
+challenge, your TLS credential provisioning system creates a DNS TXT record with
+a value expected by Let's Encrypt.
+
+Each Teleport Proxy Service instance expects TLS credentials for HTTPS to be available at the file
+paths `/etc/teleport-tls/tls.key` (private key) and `/etc/teleport-tls/tls.crt`
+(certificate).
```
@rosstimothy i'm removing this line since it isn't technically required as stated

WilliamLoy committed:
addressing review comments

dboslee commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,253 @@
+auth_service:
+ ...
+ tunnel_strategy:
+  type: proxy_peering
```
```suggestion
  type: proxy_peering
  agent_connection_count: 2
```

`agent_connection_count` is configured under tunnel_strategy in the auth config not under the `proxy_service`

WilliamLoy replied:
@dboslee good catch, I've corrected this.

WilliamLoy committed:
Update aws-gslb-proxy-peering-ha-deployment.mdx

zmb3 commented on docs/img/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.png:
I find the _User Proxy_ and _Agent Proxy_ terminology on this diagram a bit confusing.

Teleport doesn't distinguish them, they're all just proxies that handle both user and agent traffic.

WilliamLoy replied:
I copied the verbiage from the RFD for proxy peering but i'll change it.

zmb3 commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,252 @@
+This deployment architecture features two important design decisions:
+
+1. AWS Route 53 Latency based routing is used for Global Server Load Balancing 
+   [GSLB](https://www.cloudflare.com/learning/cdn/glossary/global-server-load-balancing-gslb/). 
```
Here we mention AWS and then link to a Cloudflare page. Is there an AWS GSLB doc we could link to instead?

WilliamLoy replied:
@zmb3 AWS doesn't have documentation on it. My understanding is that GSLB is an architecture type that is created through the use of DNS geographic or latency routing. It isn't a service or product actually offered by AWS.

zmb3 commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,252 @@
+clustered in a single region or for Managed Service Providers needing to provide separate clusters 
+to customers.
+
+We recommend this for globally distributed resources and end-users that prefer a single point of
```
It's not clear what _this_ is here.
```suggestion
This architecture is best suited for globally distributed resources and end-users that prefer a single point of
```

zmb3 commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,252 @@
+- High Availability Teleport Cluster
+- Deployed exclusively in AWS ecosystem
+- Auth Servers must remain in a single region
+- Proxies are deployed across multiple regions
```
```suggestion
- Proxy Servers are deployed across multiple regions
```

zmb3 commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
````diff
@@ -0,0 +1,252 @@
+
+|Record name|Type|Value|
+|---|---|---|
+|```*teleport.example.com```|CNAME|AWS Route 53 nameservers|
````
```suggestion
|```*.teleport.example.com```|CNAME|AWS Route 53 nameservers|
```

WilliamLoy committed:
Delete aws-gslb-proxy-peering-ha-deployment.png

WilliamLoy committed:
Add files via upload

WilliamLoy committed:
Update docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx

Co-authored-by: Zac Bergquist <zac.bergquist@goteleport.com>

WilliamLoy committed:
Update docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx

Co-authored-by: Zac Bergquist <zac.bergquist@goteleport.com>

WilliamLoy committed:
Update docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx

Co-authored-by: Zac Bergquist <zac.bergquist@goteleport.com>

webvictim commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,252 @@
+It is recommeded to add a wildcard record for every region if you plan to use Teleport Appplication Access.
+
+The following CNAME record values need to be set:
+- **Value:** The domain name of the NLB where example-region-1 located Teleport resource traffic should be routed
```
```suggestion
- **Value:** The domain name of the NLB where `example-region-1` located Teleport resource traffic should be routed
```

webvictim commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
````diff
@@ -0,0 +1,252 @@
+
+### Teleport resource agent configuration for GSLB
+To facilitate latency routing, resource agents  must be configured to point ```proxy_server:``` to 
+the GSLB domain configured in Route53 _not the specific proxy NLB address_.
````
Bold this to make it a little more obvious.

```suggestion
the GSLB domain configured in Route53, **not** the specific proxy NLB address.
```

webvictim commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,252 @@
+
+### Auth Service Proxy Peering configuration 
+
+The Teleport Auth Service must be configured to use the proxy_peering tunnel strategy as shown in the example below:
```
```suggestion
The Teleport Auth Service must be configured to use the `proxy_peering` tunnel strategy as shown in the example below:
```

webvictim commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,252 @@
+  type="warning"
+  title="Note"
+>
+For both Option 1 and 2, agent_connection_count on the Auth service should be set to a value >=2 to decrease
```
```suggestion
For both Options 1 and 2, `agent_connection_count` on the Auth service should be set to a value >=2 to decrease
```

webvictim commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,252 @@
+For both Option 1 and 2, agent_connection_count on the Auth service should be set to a value >=2 to decrease
+the likelihood of agents being unavailable.
+</Admonition>
+Refrence the [Proxy Service configuration](https://goteleport.com/docs/reference/config/#proxy-service) reference page 
```
```suggestion
Reference the [Proxy Service configuration](https://goteleport.com/docs/reference/config/#proxy-service) reference page 
```

webvictim commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
````diff
@@ -0,0 +1,252 @@
+  ...
+```
+
+**Option 2:** Proxies can use peer_public_addr: to advertise the proxy NLB. When using this method 
+you could incur additional latency or user experience issues because peer proxies must continually dial through the NLB until 
+they establish connection to the correct peer target.
+
+
+```
+version: v3
+teleport:
+...
+proxy_service:
+  ...
+  peer_public_addr: teleport-example-nlb-us-east-1.amazonaws.com:3021
+  ...
+```
````
IMO we shouldn't suggest that people do this for the reasons I listed above. It's easier to configure, but will result in weird behaviour on restarts or in the event of flappy connectivity and make troubleshooting harder than it needs to be (is the lack of connectivity due to a timing problem, a load balancer problem, a problem on the individual proxy host etc)

Proxy peering is not designed to constantly connect and disconnect from the LB until it gets what needs like the tunnel/agent mesh is - it's slower than that and will take a lot longer to settle into a happy state.

webvictim commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,252 @@
+
+Proxies must advertise a peer address which can be configured to use one of the two options listed below:
+
+**Option 1:** You can set peer_public_addr: to the specific name of that proxy. This is the recommended 
```
```suggestion
**Option 1:** You can set `peer_public_addr` to the specific name of that proxy. This is the recommended 
```

webvictim commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,252 @@
+...
+proxy_service:
+  ...
+  peer_public_addr: teleport.example.com:3021
```
Make it a little clearer that this is an individual proxy's hostname and not the hostname of the GSLB.

```suggestion
  peer_public_addr: teleport-proxy-eu-west-1-host1.example.com:3021
```

webvictim commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,252 @@
+   [GSLB](https://www.cloudflare.com/learning/cdn/glossary/global-server-load-balancing-gslb/). 
+   This allows for efficient distribution of traffic across resources that are globally distributed.
+
+2. Teleport's Proxy Peering is used to reduce the total number of tunnel connections in the Teleport cluster.
```
```suggestion
2. Teleport's [Proxy Peering](../architecture/proxy-peering.mdx) is used to reduce the total number of tunnel connections in the Teleport cluster.
```

webvictim commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,252 @@
+2. Teleport's Proxy Peering is used to reduce the total number of tunnel connections in the Teleport cluster.
+
+
+This deployment architecture isn’t recommended for use cases where your users or resources are 
```
nit: Don't use smart quotes

```suggestion
This deployment architecture isn't recommended for use cases where your users or resources are 
```

webvictim commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,252 @@
+- Eliminates the complexity maintaining multiple Teleport clusters across multiple regions
+- Uses the lowest latency path to connect users to resources
+- Provides a highly-resilient, redundant HA architecture for Teleport that can quickly 
+  scale with an organization’s needs.
```
nit: smart quotes

```suggestion
  scale with an organization's needs.
```

WilliamLoy committed:
Update docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx

Co-authored-by: Gus Luxton <gus@goteleport.com>

WilliamLoy committed:
Update docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx

Co-authored-by: Gus Luxton <gus@goteleport.com>

WilliamLoy committed:
Apply suggestions from code review

Co-authored-by: Gus Luxton <gus@goteleport.com>

WilliamLoy commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,252 @@
+| Port | Description |
+| - | - |
+| `443` | ALPN port for TLS Routing,  HTTPS connections to authenticate `tsh` users into the cluster, and to serve a Web UI |
+| `3021`| Proxy Peering gRPC Stream  |
```
@webvictim Instead of removing this suggestion, should we reframe it as something that is technically possible that we advise against? I'm going to guess that there will be >0 people that try to set it up this way. Or we could add on a troubleshooting section and include this as a possible misconfiguration?

WilliamLoy commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,252 @@
+  during Teleport Cluster upgrades.
+
+## Disadvantages of this deployment architecture
+- When Teleport Auth servers are limited to a single region, there is a higher likelihood 
+  of decreased availability during an AWS regional outage.
```
@webvictim like you describe at the end of your comment there are workarounds but they are typically cost-prohibitive.

WilliamLoy committed:
Update aws-gslb-proxy-peering-ha-deployment.mdx

ptgott commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
````diff
@@ -0,0 +1,246 @@
+|---|---|---|
+|```*.teleport.example.com```|CNAME|AWS Route 53 nameservers|
+
+### Teleport Proxy DNS records for GSLB:
````
```suggestion
### Teleport Proxy DNS records for GSLB:

```

ptgott commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,246 @@
+
+</Admonition>
+
+### Teleport resource agent configuration for GSLB
```
```suggestion
### Teleport resource agent configuration for GSLB

```

ptgott reviewed (commented):
It is currently not possible to discover this page from the docs site. We'll need to add:
- An entry to `docs/config.json`
- A link to the `docs/pages/deploy-a-cluster.mdx` menu page

Also, does this belong in the `docs/pages/deploy-a-cluster/deployments` section instead, which features cloud-specific deployment guides? Right now, it's in the same place within the docs as the [general "High Availability Deployments" guide](https://goteleport.com/docs/deploy-a-cluster/high-availability/). Is this a specific implementation of the general High Availability pattern on AWS? It looks like the only difference (aside from AWS-specific products) is Proxy Peering. There's a potential for confusion with the general HA guide since the two share a lot of content.

pschisa committed:
Update docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx

Co-authored-by: Paul Gottschling <paul.gottschling@goteleport.com>

pschisa commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,246 @@
+## Disadvantages of this deployment architecture
+- When Teleport Auth servers are limited to a single region, there is a higher likelihood 
+  of decreased availability during an AWS regional outage.
+- Technically complex to deploy
```
adding the proxy peering and GSLB component for multi-region low latency is more complex than single region HA, we could call out that this is in comparison to single region deployments

pschisa committed:
Apply suggestions from code review

Co-authored-by: Paul Gottschling <paul.gottschling@goteleport.com>

pschisa committed:
Update aws-gslb-proxy-peering-ha-deployment.mdx

adding clarity to technical complex disadvantage

pschisa committed:
Update aws-gslb-proxy-peering-ha-deployment.mdx

pschisa committed:
Update aws-gslb-proxy-peering-ha-deployment.mdx

pschisa committed:
Merge branch 'master' into Will/aws-gslb-proxy-peer-guide

pschisa committed:
Update aws-gslb-proxy-peering-ha-deployment.mdx

pschisa commented:
> Also, does this belong in the `docs/pages/deploy-a-cluster/deployments` section instead, which features cloud-specific deployment guides?

I think this makes sense, as the guide is specific to using AWS and not a general guide like the HA one

ptgott commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,248 @@
+This architecture is best suited for globally distributed resources and end-users that prefer a single point of
+entry while also ensuring minimal latency when accessing connected resources.
+
+## Key deployment components
```
```suggestion
## Key deployment components

```

ptgott commented on docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx:
```diff
@@ -0,0 +1,248 @@
+entry while also ensuring minimal latency when accessing connected resources.
+
+## Key deployment components
+- High-availability Teleport cluster
```
What did you think of this comment? https://github.com/gravitational/teleport/pull/24164#discussion_r1189072435

ptgott reviewed (commented):
I think all that's left is:

- [x] Putting this in `docs/pages/deploy-a-cluster/deployments`
- [x] Updating `docs/config.json` so we have a new navigation sidebar entry
- [x] Updating `docs/pages/deploy-a-cluster/deployments.mdx` and `docs/pages/deploy-a-cluster/introduction.mdx` to link to this page

pschisa committed:
Apply suggestions from code review

Co-authored-by: Paul Gottschling <paul.gottschling@goteleport.com>

pschisa committed:
Update aws-gslb-proxy-peering-ha-deployment.mdx

pschisa committed:
Update config.json

pschisa committed:
Update deployments.mdx

pschisa committed:
Update introduction.mdx

pschisa committed:
Rename docs/pages/deploy-a-cluster/aws-gslb-proxy-peering-ha-deployment.mdx to docs/pages/deploy-a-cluster/deployments/aws-gslb-proxy-peering-ha-deployment.mdx

pschisa commented:
@ptgott Completed all tasks left, please let me know what you think.  I did not move the image location as there was not a sub-directory for `deployments` already.  Let me know if the image needs to be moved for best practices

pschisa committed:
Merge branch 'master' into Will/aws-gslb-proxy-peer-guide

ptgott commented:
@pschisa Thanks for the edits! Going to push some tiny copy-edits and linter fixes, then approve.

ptgott committed:
Minor copy edits and linter fixes
